### ISSUE TITLE
[Dynamo + DDP] If DDP partitions FX graph generated by Dynamo correctly

### ISSUE THREAD

**Issue by yanboliang** [pytorch/pytorch#101624]
### 🐛 Describe the bug

Repro:
```
import torch
import logging
import torch._dynamo

torch._logging.set_logs(dynamo=logging.DEBUG, bytecode=True)

import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(420)

class MyModel(nn.Module):

    def __init__(self):
        super(MyModel, self).__init__()
        self.linear = torch.nn.Linear(5, 5)

    def forward(self, x):
        x = x + 1
        with torch.cuda.amp.autocast(dtype=torch.float16):
            x = self.linear(x)
            x = torch.sin(x)
        x = torch.cos(x)
        x = x - 1
        return x

x = torch.randn(5, 5, device="cuda")

m = MyModel().to("cuda")
print(m(x))

opt_m = torch.compile(backend="eager")(m)
print(opt_m(x))
```

Generated FX graph:
```
 ===== __compiled_fn_0 =====
 <eval_with_key>.0 class GraphModule(torch.nn.Module):
    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        
        # File: /scratch/ybliang/work/repos/debug/debug6.py:19, code: x = x + 1
        add = l_x_ + 1;  l_x_ = None
        
        # No stacktrace found for following nodes
        enter_functional_autocast = torch__dynamo_variables_ctx_manager_enter_functional_autocast('cuda', torch.float16, True, True)
        
        # File: /scratch/ybliang/work/repos/debug/debug6.py:21, code: x = self.linear(x)
        l__self___linear = self.L__self___linear(add);  add = None
        
        # File: /scratch/ybliang/work/repos/debug/debug6.py:22, code: x = torch.sin(x)
        sin = torch.sin(l__self___linear);  l__self___linear = None
        
        # No stacktrace found for following nodes
        exit_functional_autocast = torch__dynamo_variables_ctx_manager_exit_functional_autocast(enter_functional_autocast);  enter_functional_autocast = None
        
        # File: /scratch/ybliang/work/repos/debug/debug6.py:23, code: x = torch.cos(x)
        cos = torch.cos(sin);  sin = None
        
        # File: /scratch/ybliang/work/repos/debug/debug6.py:24, code: x = x - 1
        sub = cos - 1;  cos = None
        return (sub,)
```

Generated compiled function:
```
 __compiled_fn_0 <eval_with_key>.0 opcode         name                       target                                                  args                                 kwargs
-------------  -------------------------  ------------------------------------------------------  -----------------------------------  --------
placeholder    l_x_                       L_x_                                                    ()                                   {}
call_function  add                        <built-in function add>                                 (l_x_, 1)                            {}
call_function  enter_functional_autocast  <function enter_functional_autocast at 0x7f7594c66550>  ('cuda', torch.float16, True, True)  {}
call_module    l__self___linear           L__self___linear                                        (add,)                               {}
call_function  sin                        <built-in method sin of type object at 0x7f798f3f19e0>  (l__self___linear,)                  {}
call_function  exit_functional_autocast   <function exit_functional_autocast at 0x7f7594c74d30>   (enter_functional_autocast,)         {}
call_function  cos                        <built-in method cos of type object at 0x7f798f3f19e0>  (sin,)                               {}
call_function  sub                        <built-in function sub>                                 (cos, 1)                             {}
output         output                     output                                                  ((sub,),)                            {}
```

### Versions

N/A

cc @ezyang @gchanan @zou3519 @msaroufim @wconstab @ngimel @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

**Comment by yanboliang** [comment]
I'd like to know if Dynamo + DDP optimizer partitions the graph and carries the autocast state. E.g, if we partition this graph into two subgraph:
* subgraph 1: from beginning until ```l__self___linear = self.L__self___linear(add)```;
* subgraph 2: from ```sin = torch.sin(l__self___linear);``` until the end.

For subgraph 2, what is the entry autocast state? If we don't do anything, it will start from PyTorch global autocast state in the specific node, which is not correct.

**Comment by wconstab** [comment]
the ddp partitioner has no idea about the autocast enter/exit nodes, and they would end up in separate graphs in some cases.

I don't know what actually happens in that case.

If we want to support autocast by adding start/end nodes into the IR, then we will have to update any IR transformations to ensure that the start/end regions remain valid.  It seems like the autocast start/end feature got added, but without considering the DDP optimizer.

I suspect we could add support to the DDPOptimizer to patch up the autocast states.  Essentially, for each output subgraph, classify as (a) contains enter+exit, (b) contains enter only, (c) contains nothing but is after an enter, (d) contains exit only.  Then for graphs (b) we could append an exit, for graphs (c) we prepend an enter, append an exit, and graphs (d) we prepend an enter.

However, I'm not too thrilled about doing this.  At the very least, it's not great to make DDPOptimizer need to know about autocast at all.  (And this would get worse if we implement other features using similar start/end nodes). Is there a cleaner way to implement autocast?  What if it were a higher-order op?

**Comment by yanboliang** [comment]
Yes, I agree DDPOptimizer should not be aware of autocast, but we do need to figure out a way to handle this. I think all context manager states(e.g, ```torch.with_grad```, ```torch.amp.autocast```, etc) should be aware when partitioning graph, otherwise, it would cause accuracy issue.

> What if it were a higher-order op?

I'm not very clear if higher-order op suffer from this issue. cc @zou3519

**Comment by bdhirsh** [comment]
I think that making `no_grad`/`enable_grad`/`autocast` into higher order ops would technically fix this: the DDPOptimizer would see the higher order op as a single node in the graph, and wouldn't look into any of the subgraphs, so it wouldn't be possible to accidentally partition inside of one of those regions.

(cc @ezyang - we talked a bit last week about how it might be easier not to bother making e.g. `no_grad()` a higher order op, as part of pre-dispatch tracing. But doing so would likely fix this issue?)

If we decided not to make these higher order ops, then I think there are 2 other options:

(1) What Will said above - make any partitioners like DDPOptimizer aware of these context managers, and know to avoid splitting in one of these regions

(2) Have AOTAutograd handle mutations to global autograd/autocast state properly. We could in theory do something like:

(a) At trace time, reset our global state to whatever it was before the trace started. Record the changes to global state off to the side somewhere (e.g. `is_grad_enabled` changed from True to False)

(b) At runtime, after executing the compiled graph, replay the changes to global state that we recorded.


Of those 3 options (where the first option is "have dynamo convert all torch context managers into higher order ops), I'm not sure which of them is the easiest to implement. But going the higher order ops route seems reasonable.

**Comment by yanboliang** [comment]
> the DDPOptimizer would see the higher order op as a single node in the graph, and wouldn't look into any of the subgraphs, so it wouldn't be possible to accidentally partition inside of one of those regions.

I hope higher order ops can fix this, but I believe there is downside, when the context manager cover majority of the graph, then the whole graph is not splittable, and DDP optimizer can't benefit from communication/computation overlap.

For the other two options, it's either fixing at dynamo level or AOT level, both makes sense, just need to understand the complexity. I think it's not very hard in dynamo level to do that. cc @anijain2305

**Comment by wconstab** [comment]
Yea, the way Brian pitched it would not work for DDPOptimizer for the reason Yanbo mentioned.

We definitely want to analyze and split the 'flattened' graph, not the top-level graph.

Is it possible that we make more infra built into the FX partitioner infra that handles treating higher-order-ops as splittable entities and lets partitioning operate on the 'flattened' graph?
1) a util for iterating an fx graph in a 'flattened' manner, turning all higher-order-ops into metadata of the iterator (so it'd be possible when handling a particular node to see all of the higher-order-ops it was a part of, but it'd also be possible to just ignore them and treat the flattened nodes like flattened nodes)
2) the partitioner tool gets updated so when it has to assign nodes that are inside a H.O. op to different partitions, it replicates the H.O. op to contain each of the new groups of ops, and then sticks the replicated H.O. ops into the subgraphs

This would necessarily restrict what we can use Higher Order operators for, to operations that are associative, essentially.  Is that an issue?

**Comment by ezyang** [comment]
So one difficulty is that autocast is technically not associative. In particular, the autocast region maintains caching and I believe you dump the cache once you exit the region. Maybe this is not a big deal. If you want to split through higher order ops you need some way to split them, there probably is just no way around this (and maybe you just setup some APIs where you can reuse the cache on later invocations.)

One thing to note about autocast is that it happens before autograd. We need the DDP partitioner so that we can generate separate autograd custom functions to maintain pipelining. This means that, technically, we can apply autocast before we partition for DDP. So if we were willing to do a full rewrite, we could do partitioning AFTER autocast was applied. This would happen in AOTAutograd and you would have to modify AOTAutograd to be able to generate multiple custom functions.

Another thought is that, if you have a graph that is solely "change the autocast state to be X", when you run the graph, you're obligated to actually change the autocast state to X! So AOTAutograd would be responsible for tracking these global state changes and replaying them. One annoyance is that the code generated this way is not exception safe (this is what higher order ops would help with.)

**Comment by anijain2305** [comment]
If we are considering a full rewrite and have time to fix this properly, I wonder if the current work of introducing collectives at TorchInductor would obviate the need of partitioning.

**Comment by wconstab** [comment]
yea, I think we want to kill the partitioner in the long term, and have DDP be traceable into the graph.  But i'm not sure when we'll get there...

I guess we probably need to do a hack.  which hack is least bad?  Do we know how bad it would be to ignore the 'state' thing Ed mentioned and just divide one autocast region into multiple regions inside DDP?

Another idea is that we could make the partitioner refuse to break up regions between start/end.  If typical usage is fine-grained, this might not be horrible for DDP performance.  If it's really coarse, it's the same as just disabling DDPOptimizer (which we can tell users to do with a flag today)

**Comment by ezyang** [comment]
We discussed the autocast thing and we disable the cache in tracing. So autocast is fully associative and can be split.

**Comment by Chillee** [comment]
I'm going to assign this to @bdhirsh for now.

**Comment by yanboliang** [comment]
Thanks for the discussion, all makes sense!

Since this is a blocker for an important Meta internal model launch, after discussing with @wconstab and @anijain2305, we decided to have a short term solution(patch up the autocast/grad state in dynamo level) to unblock.

**Comment by anijain2305** [comment]
Closing this as the issue was resolved.
